### PR TITLE
v1.2.0

### PR DESCRIPTION
**Pull request by C-D-Lewis**
_Features_
- Support import from CSV for many simple resources with `--from-csv`.
- Support export of place resources to CSV.
- Add missing `help` command.
- Add `runCommand()` method to plugin API.

_Fixes_
- `--field` now pretty-prints output.
- Validate new Operators added with `operators add` command work before adding to config.
- Improve error output when invalid switch is specified.

### CODE REVIEW

File: src/modules/commands.js [hunk 1]
```diff
@@ -31,8 +32,8 @@ const COMMAND_LIST = [
 const showSyntax = (command) => {
   const { firstArg, operations } = command;
   const specs = Object.keys(operations).map((item) => {
-    const { pattern, buildable } = operations[item];
-    return `evrythng ${firstArg} ${pattern} ${buildable ? '(or --build)' : ''}`;
+    const { pattern, buildable, importable } = operations[item];
+    return `evrythng ${firstArg} ${pattern} ${buildable ? '(or --build)' : ''} ${importable ? '(or --from-csv)' : ''}`;
```
Comment (joshcroad) [inline src/modules/commands.js]: This output could be more inline with standard cli tools, such as:
```bash
evrythng thngs create [$payload|--build|--from-csv]
...
```

Comment (joshcroad) [reply src/modules/commands.js]: This change could be made in the commands themselves, like thngs:
```
printPattern: 'create [$payload|--build|--from-csv]',
```

File: src/modules/http.js [hunk 2]
```diff
@@ -187,13 +187,13 @@ const printResponse = async (res) => {
   // Get just one field
   const field = switches.FIELD;
   if (field) {
-    logger.info(data[field]);
+    logger.info(JSON.stringify(data[field], null, 2));
```
Comment (joshcroad) [inline src/modules/http.js]: Can this `2` be set as a constant at the top?

File: tests/commands/operator.js [hunk 3]
```diff
@@ -61,4 +66,10 @@ describe('operators', () => {
   it('should not throw error for \'operators $name remove\'', async () => {
     return cli(`operators ${ctx.operatorName} remove`).should.be.fulfilled;
   });
+
+  it('should throw error if operator credentials are not real', async () => {
+    sinon.restore();
+
+    return operator.validateCredentials('us', 'somebadapikey').should.be.rejected;
```
Comment (joshcroad) [inline tests/commands/operator.js]: You _should_ use expect

File: tests/modules/csv.js [hunk 4]
```diff
@@ -4,37 +4,129 @@
  */
 
 const { expect } = require('chai');
+const _ = require('lodash');
```
Comment (joshcroad) [inline tests/modules/csv.js]:
```
const { isEqual } = require('lodash');
```

File: tests/modules/csv.js [hunk 5]
```diff
@@ -4,37 +4,129 @@
...
+    const obj = { foo: 'bar', 'baz': 'thng' };
+    const expected = '{foo:bar|baz:thng}';
+    expect(csv.encodeObject(obj)).to.equal(expected);    
+  });
+
+  it('should decode a simple JSON object encoded string', () => {
+    const objStr = '{foo:bar|baz:thng}';
+    const expected = { foo: 'bar', 'baz': 'thng' };
+    const result = csv.decodeObject(objStr);
+    expect(_.isEqual(result, expected)).to.equal(true);
```
Comment (joshcroad) [inline tests/modules/csv.js]: Add tests for preserving types, such as boolean and number

File: src/modules/csv.js [hunk 6]
```diff
@@ -3,70 +3,394 @@
...
+      res[key] = {
+        type: 'Point',
+        coordinates: [parseFloat(lon), parseFloat(lat)],
+      };
+      return res;
+    }
+
+    // Simple value
+    res[key] = value;
+    return res;
```
Comment (joshcroad) [inline src/modules/csv.js]: This function converts unmatched values to string only, ⚠️JSON conversions should include boolean and number (including float)

Comment (C-D-Lewis) [reply src/modules/csv.js]: Type is inferred now at the top of the function.

File: src/modules/csv.js [hunk 7]
```diff
@@ -3,70 +3,412 @@
...
     Object.keys(arrItem)
-      .filter(itemKey => !result.includes(itemKey))
+      .filter(itemKey => !res.includes(itemKey))
       .filter(itemKey => !IGNORE.includes(itemKey))
-      .forEach(newItem => result.push(newItem));
-  });
-
-  const buildKey = item => prefix ? `${prefix}.${item}` : item;
-  return result.map(buildKey);
+      .forEach(newItem => res.push(newItem));
```
Comment (joshcroad) [inline src/modules/csv.js]: We can half the complexity of this function
```js
return arr.reduce((res, arrItem) => res.concat(Object.keys(arrItem)
    .filter(itemKey => !res.includes(itemKey) && !IGNORE.includes(itemKey)), []);
```

File: src/modules/csv.js [hunk 8]
```diff
@@ -3,70 +3,412 @@
...
+  ].join(',');
 
   const rows = arr.map((item) => {
-    const { customFields, identifiers } = item;
-    return [
-      addCells(item, objKeys),
-      addCells(customFields, cfKeys),
-      addCells(identifiers, idKeys)
-    ].join(',');
+    return objectToCells(item, object)
```
Comment (joshcroad) [inline src/modules/csv.js]: No need to `return` here - simply return the function call
```js
const rows = arr.map(item => objectToCells(item, object)
  ...
```

File: src/modules/csv.js [hunk 9]
```diff
@@ -3,70 +3,412 @@
...
+};
+
+/**
+ * Preserve the type of a number or boolean that was extracted as a string token.
+ *
+ * @param {string} value - The string to process.
+ * @returns {string|number|boolean} The extracted value in its 'original' type.
+ */
+const preserveType = (value) => {
+  if (!isNaN(Number(value))) {
```
Comment (joshcroad) [inline src/modules/csv.js]: This is quite unreliable

File: src/modules/csv.js [hunk 10]
```diff
@@ -3,70 +3,412 @@
...
+  }
+
+  // Get headers
+  const csvStr = fs.readFileSync(path, 'utf8').toString();
+  const rows = await neatCsv(csvStr);
+
+  const scope = new evrythng.Operator(operator.getKey());
+  await util.nextTask(rows.map(item => () => {
+    const resource = rowToObject(item);
+    return createResource(scope, resource, type);
```
Comment (joshcroad) [inline src/modules/csv.js]: One-liner here

**Pull request merged**